### PR TITLE
Implement RFC 51: Add `ShapeCastable.from_bits` and `amaranth.lib.data.Const`.

### PR DESCRIPTION
**Pull request by wanda-phi**
TODO:

- [x] tests
- [x] documentation
- [ ] ~~add custom `Const` subclasses for `Struct`/`Union`?~~ deferred

### CODE REVIEW

File: amaranth/lib/data.py [hunk 1]
```diff
@@ -590,7 +594,7 @@ class View(ValueCastable):
     an Amaranth value instead of a constant integer. The returned element is chosen dynamically
     in that case.
 
-    A view can only be compared for equality with another view of the same layout,
+    A view can only be compared for equality with another view or const of the same layout,
```
Comment (whitequark) [inline amaranth/lib/data.py:597]:
```suggestion
    A view can only be compared for equality with another view or constant of the same layout,
```

File: amaranth/lib/data.py [hunk 2]
```diff
@@ -789,6 +799,221 @@ def __repr__(self):
...
+    """A constant value viewed through the lens of a layout.
+
+    The :class:`Const` class is a constant analog of :class:`View`. It implements largely
+    the same operations, except that it deals with compile-time constants instead of arbitrary
+    :class:`Value`.
+
+    Creating a const
+    ################
+
+    A constant can be created from a dict or list of field values using :meth:`Layout.const`,
```
Comment (whitequark) [inline amaranth/lib/data.py:812]: I add links to the Python stdlib just to be completely unambiguous on what we mean at all times.

```suggestion
    A constant can be created from a :class:`dict` or :class:`list` of field values using :meth:`Layout.const`,
```

File: amaranth/lib/data.py [hunk 3]
```diff
@@ -789,6 +799,221 @@ def __repr__(self):
...
+    or from a bit-pattern using :meth:`Layout.from_bits`.
+
+    Accessing a const
+    #################
+
+    Slicing a const or accessing its attributes returns a part of the underlying value
+    corresponding
+    to the field with that index or name. If the shape of the field is a :class:`Layout`,
+    the accessed value will be another :class:`Const`. If it is another :ref:`shape-like <lang-shapelike>`
+    object implementing ``from_bits``, it will be the result of calling that method. Otherwise,
```
Comment (whitequark) [inline amaranth/lib/data.py:822]: Should be a link to ShapeCastable.from_bits

File: amaranth/lib/data.py [hunk 4]
```diff
@@ -789,6 +799,221 @@ def __repr__(self):
...
+    it will be a plain :class:`int`.
+
+    Slicing a view whose layout is an :class:`ArrayLayout` can be done with an index that is
+    an Amaranth value instead of a constant integer. The returned element is chosen dynamically
+    in that case, and the resulting value will be a :class:`View` instead of :class:`Const`.
```
Comment (whitequark) [inline amaranth/lib/data.py:827]: Oh, that's... a really fascinating behavior I haven't thought about at all. No issue with it, just that it's kind of unusual.

File: amaranth/lib/data.py [hunk 5]
```diff
@@ -789,6 +799,221 @@ def __repr__(self):
...
+
+    A const can be compared for equality with another const or view of the same layout.
+    When compared with another const, the result will be a Python :class:`bool`.
+    When compared with a view, the result will be a single-bit value. No other operators
```
Comment (whitequark) [inline amaranth/lib/data.py:831]: Here the disambiguation is actually important!

```suggestion
    When compared with a view, the result will be a single-bit :class:`Value`. No other operators
```

File: amaranth/lib/data.py [hunk 6]
```diff
@@ -789,6 +799,221 @@ def __repr__(self):
...
+        """Convert to Amaranth value.
+
+        Returns
+        -------
+        :class:`amaranth.hdl.Const`
+            A constant Amaranth value with the underlying bit-pattern.
+        """
+        return hdl.Const(self.__target, self.__layout.size)
+
+    def as_bits(self):
```
Comment (whitequark) [inline amaranth/lib/data.py:876]: Maybe add a warning to `View` that using a field called `as_bits` will break `Const`?

Comment (wanda-phi) [reply amaranth/lib/data.py:876]: It ended up a little awkward (I think the documentation will read better when we have custom `Const` subclass support?)

**Review by whitequark — APPROVED** [review]
LGTM

**Pull request merged**